tests/integration/test_integration.py

```diff
@@ -470,6 +472,74 @@ def test_kinesis_lambda_forward_chain(self):
         kinesis.delete_stream(StreamName=TEST_CHAIN_STREAM1_NAME)
         kinesis.delete_stream(StreamName=TEST_CHAIN_STREAM2_NAME)
 
+    def test_sqs_batch_lambda_forward(self):
+        sqs = aws_stack.connect_to_service('sqs')
+        lambda_api = aws_stack.connect_to_service('lambda')
+
+        # deploy test lambda connected to SQS queue
+        sqs_queue_info = testutil.create_sqs_queue(TEST_LAMBDA_NAME_QUEUE)
+        queue_url = sqs_queue_info['QueueUrl']
+        zip_file = testutil.create_lambda_archive(load_file(TEST_LAMBDA_PYTHON), get_content=True,
+            libs=TEST_LAMBDA_LIBS, runtime=LAMBDA_RUNTIME_PYTHON27)
+        resp = testutil.create_lambda_function(
+            func_name=TEST_LAMBDA_NAME_QUEUE,
+            zip_file=zip_file,
+            event_source_arn=sqs_queue_info['QueueArn'],
+            runtime=LAMBDA_RUNTIME_PYTHON27
+        )
+
+        event_source_id = resp['CreateEventSourceMappingResponse']['UUID']
+        lambda_api.update_event_source_mapping(
+            UUID=event_source_id,
+            BatchSize=5
+        )
+
+        messages_to_send = [
+            {
+                'Id': 'message{:02d}'.format(i),
+                'MessageBody': 'msgBody{:02d}'.format(i),
+                'MessageAttributes': {
+                    'CustomAttribute': {
+                        'DataType': 'String',
+                        'StringValue': 'CustomAttributeValue{:02d}'.format(i)
+                    }
+                }
+            }
+            for i in range(1, 22)
+        ]
+
+        initial_invocation_count = get_lambda_invocations_count(TEST_LAMBDA_NAME_QUEUE)
+
+        # send 21 messages (which should get split into 5 batches)
+        sqs.send_message_batch(QueueUrl=queue_url, Entries=messages_to_send[:10])
+        sqs.send_message_batch(QueueUrl=queue_url, Entries=messages_to_send[10:20])
+        sqs.send_message(
+            QueueUrl=queue_url,
+            MessageBody=messages_to_send[20]['MessageBody'],
+            MessageAttributes=messages_to_send[20]['MessageAttributes']
+        )
+
+        def wait_for_empty():
+            resp = sqs.get_queue_attributes(
+                QueueUrl=queue_url,
+                AttributeNames=['ApproximateNumberOfMessages'],
+            )
+            count = int(resp['Attributes'].get('ApproximateNumberOfMessages'))
+            if count > 0:
+                raise ValueError('expecting empty queue, has {} messages'.format(count))
+            return count
+
+        # wait for the queue to drain
+        retry(wait_for_empty, retries=20, sleep=0.5)
+        # give additional time for the lambdas to finish running
+        time.sleep(2.0)
+
+        # batch size is 5, so 21 messages gets sent in 5 batches
+        final_invocation_count = get_lambda_invocations_count(TEST_LAMBDA_NAME_QUEUE)
+        invocation_count = final_invocation_count - initial_invocation_count
+        self.assertEqual(invocation_count, 5)
```

Comment:
Looks like we're currently still seeing an error in the CI builds, which could be related to a timing issue here. 

```
======================================================================
FAIL: test_sqs_batch_lambda_forward (tests.integration.test_integration.IntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/localstack/localstack/tests/integration/test_integration.py", line 540, in test_sqs_batch_lambda_forward
    self.assertEqual(invocation_count, 5)
AssertionError: 3.0 != 5
-------------------- >> begin captured logging << --------------------
```

Perhaps we should remove the `time.sleep(2.0)` above, and add the assertion in line 540 also to the retry function (`wait_for_empty` - probably could be renamed to `wait_for_done` in this case).

Reply:
Ok, hopefully my edit should have just resolved the CI timing issue. It actually also let me improve the test a bit. Instead of just checking that the lambda was invoked a certain number of times, it also now asserts that all of the messages were deleted from the queue.

I didn't realize it before, but the test lambda function it was using actually requires a kinesis stream to exist so it can forward events, so I switched it to the echo lambda since all I care about here is that the lambda was invoked and returned a valid response.

It no longer has a static sleep period defined but will check for the queue to be empty (no pending messages, no delayed messages, no invisible messages) as well as check for the correct number of lambda invocations every 3s for 10 times.

Reply:
ok, @whummer, I resolved a couple of remaining issues causing the CI checks to fail, including a change in the `get_lambda_metrics` utility function. If the changes look good to you I'll go ahead and squash + push again